File: src/main/java/com/parmet/squashlambdas/cal/Action.java
package com.parmet.squashlambdas.cal;

import com.google.common.collect.ImmutableList;
import com.parmet.squashlambdas.activity.Activity;
import java.io.IOException;
import java.util.Collection;

public enum Action {
  CREATE {
    @Override
    public void handle(Activity activity, EventManager manager) throws IOException {
      manager.create(activity);
    }
  },
  UPDATE {
    @Override
    public void handle(Activity activity, EventManager manager) throws IOException {
      manager.update(activity);
    }
  },
  DELETE {
    @Override
    public void handle(Activity activity, EventManager manager) throws IOException {
      manager.delete(activity);
    }
  },
  NONE {
    @Override
    public void handle(Activity activity, EventManager manager) {
    }
  };

  private static final ImmutableList<String> CREATION =
      ImmutableList.of(
          "A reservation including you has been made",
          "You've joined a reservation",
          "You have been added to the activity");

  private static final ImmutableList<String> UPDATING =
      ImmutableList.of(
          "has been removed",
          "has joined your reservation");

  private static final ImmutableList<String> DELETION =
      ImmutableList.of(
          "You've been removed from a reservation",
          "You have successfully cancelled a reservation",
          "A reservation including you has been cancelled");

  private static final ImmutableList<String> NO_ACTION =
      ImmutableList.of(
          "This is a reminder",
          "Here are your scores recorded",
          "has cancelled your reservation",
          "has re-confirmed a reservation made");

  public static Action parseFromSubject(String body) {
    if (containsMatch(body, CREATION)) {
      return CREATE;
    }
    if (containsMatch(body, UPDATING)) {
      return UPDATE;
    }
    if (containsMatch(body, DELETION)) {
      return DELETE;
    }
    if (containsMatch(body, NO_ACTION)) {
      return NONE;
    }

    throw new IllegalArgumentException("Unable to parse action from " + body);
  }

  private static boolean containsMatch(String body, Collection<String> matches) {
    return matches.stream().anyMatch(body::contains);
  }

  public abstract void handle(Activity activity, EventManager manager) throws IOException;
}

File: src/main/kotlin/com/parmet/squashlambdas/cal/Action.kt
package com.parmet.squashlambdas.cal

import com.parmet.squashlambdas.activity.Activity

sealed class Action(
    @Transient
    internal val handle: (Activity, EventManager) -> Unit
) {
    object Create : Action({ activity, manager -> manager.create(activity) })

    object Update : Action({ activity, manager -> manager.update(activity) })

    object Delete : Action({ activity, manager -> manager.delete(activity) })

    object None : Action({ _, _ -> })

    override fun toString() = this::class.java.simpleName

    companion object {
        private val CREATION =
            listOf(
                "A reservation including you has been made",
                "You've joined a reservation",
                "You have been added to the activity"
            )

        private val UPDATING =
            listOf(
                "has cancelled out of reservation",
                "has been removed",
                "has joined your reservation"
            )

        private val DELETION =
            listOf(
                "You've been removed from a reservation",
                "has been successfully cancelled",
                "has cancelled your reservation",
                "A reservation including you has been cancelled",
                "You have successfully cancelled"
            )

        private val NO_ACTION =
            listOf(
                "This is a reminder",
                "Here are your scores recorded",
                "has re-confirmed a reservation made"
            )

        fun parseFromSubject(body: String): Action {
            return when {
                CREATION.containsMatch(body) -> Create
                UPDATING.containsMatch(body) -> Update
                DELETION.containsMatch(body) -> Delete
                NO_ACTION.containsMatch(body) -> None
                else -> throw IllegalArgumentException("Unable to parse action from $body")
            }
        }

        private fun List<String>.containsMatch(body: String) = any { body.contains(it) }
    }
}
